llap-common/src/java/org/apache/hadoop/hive/llap/AsyncResponseHandler.java

```diff
@@ -0,0 +1,114 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncResponseHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncResponseHandler.class);
+
+  private final AtomicBoolean isShutdown = new AtomicBoolean(false);
+
+  private final AsyncPbRpcProxy.RequestManager requestManager;
+  private final ExecutorService responseWaitingService = Executors.newSingleThreadExecutor();
+  private final LinkedBlockingDeque<AsyncCallableRequest<Message, Message>>
+          incomingResponseFutures = new LinkedBlockingDeque<>();
+
+  public AsyncResponseHandler(AsyncPbRpcProxy.RequestManager requestManager) {
+    this.requestManager = requestManager;
+  }
+
+  public void start() {
+    responseWaitingService.submit(new AsyncResponseHandlerRunnable());
+  }
+
+  public void addToAsyncResponseFutureQueue(AsyncCallableRequest<Message, Message> request) {
+    incomingResponseFutures.add(request);
+  }
+
+  public void shutdownNow() {
+    isShutdown.set(true);
+    responseWaitingService.shutdownNow();
+  }
+
+  private final class AsyncResponseHandlerRunnable implements Runnable {
+
+    private final List<AsyncCallableRequest<Message, Message>> responseFuturesQueue = new ArrayList<>();
+
+    @Override
+    public void run() {
+      while (!isShutdown.get()) {
+        Iterator<AsyncCallableRequest<Message, Message>> iterator = responseFuturesQueue.iterator();
+        while (iterator.hasNext()) {
+          AsyncCallableRequest<Message, Message> request = iterator.next();
+          AsyncGet<Message, Exception> responseFuture = request.getResponseFuture();
+          if (responseFuture != null && responseFuture.isDone()) {
+            try {
+              iterator.remove();
+              LlapNodeId nodeId = request.getNodeId();
+              // since isDone is true, getFuture.get should return immediately
+              try {
+                Message remoteValue = responseFuture.get(-1, TimeUnit.MILLISECONDS);
+                if (remoteValue instanceof Throwable) {
+                  request.getCallback().indicateError((Throwable) remoteValue);
+                } else {
+                  request.getCallback().setResponse(remoteValue);
+                }
+              } catch (Exception e) {
+                request.getCallback().indicateError(e);
+              } finally {
+                requestManager.requestFinished(nodeId);
+              }
+            } catch (Throwable e) {
+              LOG.warn("ResponseDispatcher caught", e);
+            }
+          }
+        }
+        try {
+          if (responseFuturesQueue.isEmpty()) {
+            // there are no more futures to hear from, just block on incoming futures
+            AsyncCallableRequest<Message, Message> request = incomingResponseFutures.poll(200, TimeUnit.MILLISECONDS);
```

Comment:
Instead of 200 ms polling, can you check with wait/notify or blocking DS? (LinkedBlockingDeque can be tried out with sync).